### ISSUE TITLE
Consider adding "ident" service to TLS tunnel

### ISSUE THREAD

**Issue by vanrein** [arpa2/tlspool#52]
The [ident protocol](https://tools.ietf.org/html/rfc1413) listens on TCP port 113 (decimal) of the same IP as to which the TLS tunnel binds (or from which it connects).  It responds to inquiries about identity; we might use a non-standard format,
```
6193, 80 : USERID : ARPA2 : bakker@orvelte.nep
```
this could help us get started with web service, without having created modules for all the web servers.  The underlying assumption is of course that the web server receives traffic that has TLS unwrapped.

A few alternative example error replies might be
```
6193, 80 : ERROR : NO-USER
6193, 80 : ERROR : INVALID-PORT
```

Note that the RFC permits reading multiple queries over the same TCP link.  This is quite attractive for the efficiency of a local roll-out!  Now to find a way to pool these connections...

**Comment by vanrein** [comment]
The [environment variables](http://www.cgi101.com/class/ch3/text.html) to use are `REMOTE_ADDR` and `REMOTE_PORT`, and after querying the identity, `REMOTE_USER` can be set.  If no user identity or an empty one is found, `REMOTE_USER` can be reset.

**Comment by vanrein** [comment]
An alternative design, specifically for use with a TLS Tunnel running on localhost, could be based on [shared memory](http://pubs.opengroup.org/onlinepubs/009695399/functions/shmget.html).

The name of a memory segment would be named with the help of [ftok()](http://pubs.opengroup.org/onlinepubs/9699919799/functions/ftok.html) and that can be used to share between web apps and TLS Tunnel.

The size of the memory is 65536 * 128, so the available ports on `::1`; a newly found connection is set to all NUL characters, and after successful `tlspool_starttls()`, the value is replaced with the returned remote identity.

Only in server mode is this of any use.

**Comment by vanrein** [comment]
Yet another variant would be to use [memcached](http://www.memcached.org).  The solution based on shared memory would be faster but constrained to localhost; the ident-based solution would be comparable in speed to memcache.  A reasonable timeout (such as 30 seconds) can be set to accommodate that we don't see the connection being closed.

A new connection would be setup before forwarding the connection to the backend.  This means that the TLS Tunnel first does a `bind()` call to know its outgoing port, before it calls `connect()`.

**Comment by vanrein** [comment]
Apache runs a [separate script](https://code.google.com/archive/p/modwsgi/wikis/ConfigurationDirectives.wiki) for authentication, namely `WSGIAuthUserScript`.

Nginx includes WSGI [through uwsgi](http://uwsgi-docs.readthedocs.io/en/latest/Nginx.html) -- no idea how to process that...

**Comment by vanrein** [comment]
**Problem:** The information from SNI would normally have to be shared too, so at least a check can be made against the "inner" hostname.  This means that the results start to look different (and not quite like a client userid at all).  How far should we drift from the intentions of the `ident` protocol?

**Comment by vanrein** [comment]
This approach is not of the same quality as #60, so this issue is closed to give #60 all the room it needs.
